doc/md: sequelize composite types (#2975)
ariga/atlas: doc/md/guides/orms/sequelize/composite-types.mdx

and may be used multiple times in different columns and models.

In order to extend our PostgreSQL schema to include both custom composite types and our Sequelize types, we configure Atlas to
read the state of the schema from a [Composite Schema](/atlas-schema/projects#data-source-composite_schema)
data source. Follow the steps below to configure this for your project:

1\. Create a `schema.sql` that defines the necessary composite type. In the same way, you can configure the composite type in
[Atlas Schema HCL language](/atlas-schema/hcl-types#composite-type):

<Tabs>
<TabItem value={"sql"} label={"Using SQL"}>

```sql title="schema.sql"
CREATE TYPE address AS (
  street text,
  city   text
);
```

</TabItem>
<TabItem value={"hcl"} label={"Using HCL"}>

```hcl title="schema.hcl"
schema "public" {}

composite "address" {
  schema = schema.public
  field "street" {
    type = text
  }
  field "city" {
    type = text
  }
}
```

</TabItem>
</Tabs>

2\. In your Sequelize models, define a column that uses the composite type only in PostgreSQL dialect:

```javascript title="user.js"
'use strict';
module.exports = function(sequelize, DataTypes) {
    const Address = {

>         type: 'public.Address', // Replace with your schema name if different

Reviewer: ```suggestion
        type: 'address', // Replace with your schema name if different
```

[Atlas Schema HCL language](/atlas-schema/hcl-types#composite-type):

<Tabs>
<TabItem value={"sql"} label={"Using SQL"}>

```sql title="schema.sql"
CREATE TYPE address AS (
  street text,
  city   text
);
```

</TabItem>
<TabItem value={"hcl"} label={"Using HCL"}>

```hcl title="schema.hcl"
schema "public" {}

composite "address" {
  schema = schema.public
  field "street" {
    type = text
  }
  field "city" {
    type = text
  }
}
```

</TabItem>
</Tabs>

2\. In your Sequelize models, define a column that uses the composite type only in PostgreSQL dialect:

<Tabs>
<TabItem value={"ts"} label ={"TypeScript"}>
<Tabs>
<TabItem value={"user"} label={"user.ts"}>

```typescript title="user.ts"
import { Table, Column, Model, PrimaryKey, AutoIncrement, DataType, AllowNull, HasOne } from "sequelize-typescript";
import Address from "./address";

@Table({ tableName: "Users" })
class User extends Model {
    @PrimaryKey
    @AutoIncrement
    @Column(DataType.INTEGER)
    id!: number;

    @AllowNull(false)
    @Column
    name!: string;

    @HasOne(() => Address)
    address!: Address;
}

export default User;
```

</TabItem>
<TabItem value={"address"} label={"address.ts"}>

> ```typescript title="address.ts"
> import { Table, Column, Model, PrimaryKey,ForeignKey, AutoIncrement, DataType, BelongsTo, AllowNull} from "sequelize-typescript";
> import User from "./user";
> 
> @Table({ tableName: "Addresses" })
> class Address extends Model {
>     @PrimaryKey
>     @AutoIncrement
>     @Column(DataType.INTEGER)
>     id!: number;
> 
>     @AllowNull(false)
>     @Column
>     street!: string;
> 
>     @AllowNull(false)
>     @Column
>     city!: string;
> 
>     @AllowNull(false)
>     @ForeignKey(() => User)
>     @Column(DataType.INTEGER)
>     userID!: number;
> 
>     @BelongsTo(() => User)
>     user!: User;
> }
> 
> export default Address;
> ```

Reviewer: ```suggestion
```typescript title="address.ts"
import { Table, Column, Model, PrimaryKey,ForeignKey, AutoIncrement, DataType, BelongsTo, AllowNull} from "sequelize-typescript";
import User from "./user";

@Table({ tableName: "Addresses" })
class Address extends Model {
    @PrimaryKey
    @AutoIncrement
    @Column(DataType.INTEGER)
    id!: number;

    @AllowNull(false)
    @Column
    street!: string;

    @AllowNull(false)
    @Column
    city!: string;

    @AllowNull(false)
    @ForeignKey(() => User)
    @Column(DataType.INTEGER)
    userID!: number;

    @BelongsTo(() => User)
    user!: User;
}

export default Address;
```
```
delete

    @AllowNull(false)
    @Column
    name!: string;

    @HasOne(() => Address)
    address!: Address;
}

export default User;
```

</TabItem>
<TabItem value={"address"} label={"address.ts"}>

```typescript title="address.ts"
import { Table, Column, Model, PrimaryKey,ForeignKey, AutoIncrement, DataType, BelongsTo, AllowNull} from "sequelize-typescript";
import User from "./user";

@Table({ tableName: "Addresses" })
class Address extends Model {
    @PrimaryKey
    @AutoIncrement
    @Column(DataType.INTEGER)
    id!: number;

    @AllowNull(false)
    @Column
    street!: string;

    @AllowNull(false)
    @Column
    city!: string;

    @AllowNull(false)
    @ForeignKey(() => User)
    @Column(DataType.INTEGER)
    userID!: number;

    @BelongsTo(() => User)
    user!: User;
}

export default Address;
```

</TabItem>
</Tabs>

</TabItem>
<TabItem value={"js"} label={"JavaScript"}>
<Tabs>
<TabItem value={"user"} label={"user.js"}>

```javascript title="user.js"
'use strict';
module.exports = function(sequelize, DataTypes) {
    const User = sequelize.define('User', {
        name: {
            type: DataTypes.STRING,
            allowNull: false
        },
    });

> 

Reviewer: ```suggestion
```

    User.associate = (models) => {
        User.hasOne(models.Address, {
            foreignKey: {
                name: 'userID',
                allowNull: false
            },
            as: 'addresses'
        });
    };

> 

Reviewer: ```suggestion
```

    return User;
};
```

</TabItem>
<TabItem value={"address"} label={"address.js"}>

```javascript title="address.js"
'use strict';
module.exports = (sequelize, DataTypes) => {
    const Address = sequelize.define('Address', {
        street: {
            type: DataTypes.STRING,
            defaultValue: false,
        },
        city: {
            type: DataTypes.STRING,
            defaultValue: false,
        },
    });

> 
>     Address.associate = (models) => {
>         Address.belongsTo(models.User, {
>             foreignKey: {
>                 name: 'userID',
>                 allowNull: false
>             },
>             as: 'addresses'
>         });
>     };
> 

Reviewer: ```suggestion
    Address.associate = (models) => {
        Address.belongsTo(models.User, {
            foreignKey: {
                name: 'userID',
                allowNull: false
            },
            as: 'addresses'
        });
    };
```
